Objects/enumobject.c

```diff
@@ -429,20 +429,21 @@ static PyObject *
 reversed_next(reversedobject *ro)
 {
     PyObject *item;
-    Py_ssize_t index = ro->index;
+    Py_ssize_t index = _Py_atomic_add_ssize(&(ro->index), -1);
 
     if (index >= 0) {
         item = PySequence_GetItem(ro->seq, index);
         if (item != NULL) {
-            ro->index--;
             return item;
         }
         if (PyErr_ExceptionMatches(PyExc_IndexError) ||
             PyErr_ExceptionMatches(PyExc_StopIteration))
             PyErr_Clear();
     }
-    ro->index = -1;
+    FT_ATOMIC_STORE_SSIZE_RELAXED(ro->index, -1);
+#ifndef Py_GIL_DISABLED
```

Comment:
I didn't take a look deeply but is it safe not to call Py_Clear at free-threded build?

Reply:
If we clear `ro->seq` in one thread, then in another thread we could be using `ro->seq` where the underlying sequence has already been deallocated. That can happen in `reversed_next`, but also in `reversed_len` and probably some other methods. 

I managed to crash the interpreter without this change (script below), although I am not 100% certain it is due to this since there is no traceback.
<details>
```
from threading import Thread
import time

def test_reversed_threading(self):
    # Test that when reading out with multiple treads no identical elements are returned
    def work(r, output):
        while True:
            try:
                l = r.__length_hint__()
                value = next(r)
            except StopIteration:
                break
            else:
                output.append(value)
    number_of_threads = 14
    x = tuple(range(12_000))
    r = reversed(x)
    worker_threads = []
    output = []
    for ii in range(number_of_threads):
        worker_threads.append(Thread(target=work, args=[r, output]))
    _ = [t.start() for t in worker_threads]
    _ = [t.join() for t in worker_threads]
    
    if sorted(output) != sorted(x):
        print( set(output).difference(set(x)),set(x).difference(set(output)) )
        if len(x)<40:
            print(f'{len(output)=}, {len(x)=}')
           # print(x)
            print(f'{sorted(output)=}')
            print('---')
            print(output)
        raise ValueError('reversed returned value from sequence twice')
    print(sorted(output))

t0=time.time()
test_reversed_threading(None)
dt=time.time()-t0
print(f'Done: {dt:.2f}')
```
</details>

Reply:
I am asking from the view of memory leakage.

Reply:
Yeah, we should not call `Py_CLEAR(ro->seq)` in the free-threaded build. Let it get freed when the iterator is deallocated. We do something similar in the other iterators like `listiter_next`.

Reply:
We also need to update the check in `reversed_len` to handle `ro->index == -1`, because `ro->seq == NULL` won't always indicate an exhausted iterator.